Diffing_Engine/Compute/Diffing.cs

```diff
@@ -0,0 +1,168 @@
+﻿using BH.oM.Base;
+using BH.Engine;
+using BH.oM.Data.Collections;
+using BH.oM.Diffing;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Security.Cryptography;
+using System.Reflection;
+using BH.Engine.Serialiser;
+using System.ComponentModel;
+using BH.oM.Reflection.Attributes;
+using BH.oM.Testing;
+
+namespace Diffing_Engine
+{
+    public static partial class Compute
+    {
+        ///***************************************************/
+        ///**** Public Methods                            ****/
+        ///***************************************************/
+        /// Diffing requires to memorize the hash of the object somewhere.
+        /// If using BHoMObjects, we can save the hash inside the objects themselves.
+        /// Otherwise another support is needed (table with | object | Hash | )
+
+        public static Delta Diffing(string projectName, List<IBHoMObject> CurrentObjs, List<string> exceptions = null, bool useDefaultExceptions = true)
+        {
+            // Clone the current objects
+            List<IBHoMObject> CurrentObjs_cloned = CurrentObjs.Select(obj => BH.Engine.Base.Query.DeepClone(obj)).ToList();
+
+            // Create project fragment
+            DiffProjFragment diffProj = new DiffProjFragment(projectName);
+
+            if (useDefaultExceptions)
+                SetDefaultExceptions(ref exceptions);
+
+            // Calculate and set the object hashes
+            CurrentObjs_cloned.ForEach(obj =>
+                obj.Fragments.Add(
+                    new DiffHashFragment(Compute.SHA256Hash(obj, exceptions), diffProj)
+                    ));
+
+            return new Delta(diffProj, CurrentObjs_cloned);
+        }
+
+        public static Delta Diffing(List<IBHoMObject> CurrentObjs, List<IBHoMObject> ReadObjs, bool propertyLevelDiffing = true, List<string> exceptions = null, bool useDefaultExceptions = true)
```

Comment:
> Another point is, why would anyone avoid using `propertyLevelDiffing`?

Because no matter how efficient, PropertyLevelDiffing is bound to be slower than the collection level diffing. There will be cases where the propertyLevelDiffing will be very slow to use while it might be enough for someone to know what's changed at the collection level only. Since it's just a boolean that can be defaulted, I think it's important to offer this option from the outset.

Reply:
> I'd make this much more general, i.e. `Diffing(List<IBHoMObject> setA, List<IBHoMObject> setB`.

While I agree with the naming (`setA`, `setB`) I'm not sure I understand what you mean as "general" here. 
The method is already as generic as possible I think -- the inclusion of the other 2 parameters is entirely optional as they're defaulted, but I always felt they have an important role because they allow the user to define how the diffing should be done. 
In particular, we started drafting the diffing with Al to use comparers, and we retain the flexibility of the comparers by allowing the user to specify `exceptions` in the Hash definition here.

Reply:
> I also don't see a big difference with the method above.

> It is extremely general, and the case above is just a specific case where `CurrentObjs` is an empty list.

Perhaps you meant to say _where **`readObjs`** is an empty list_?

Here I kind of agree, I separated the two methods like that because my idea was to expose the first method when creating a brand new diffing project, and the following one when dealing with an existing diffing project. 

I was thinking to have a different, clearer way to create a bran new diffing project though, I will work on it a bit.

Reply:
I meant that it might not be a temporal difference, or a difference between something considered `current` and `read`. It may just be I created two objects, and I want to know the difference between them.

Probably `left` and `right can be another option?



> Perhaps you meant to say _where **`readObjs`** is an empty list_?
> 
> Here I kind of agree, I separated the two methods like that because my idea was to expose the first method when creating a brand new diffing project, and the following one when dealing with an existing diffing project.

Yes, exactly. From a UX point of view, I think that it doesn't matter if the project already exist or not. I just want to push to it.

We can then have another method `Create.DiffingProject` that might create an empty project. But I would avoid to put it into a `Compute.Diff`

Reply:
> I meant that it might not be a temporal difference, or a difference between something considered current and read. It may just be I created two objects, and I want to know the difference between them.
> 
> Probably left and `right can be another option?

Oh sorry, I thought you meant the inclusion of the the other input parameters! Then I agree thoroughly, and I think that `SetA` and `SetB` are already a good nomenclature.

> We can then have another method `Create.DiffingProject` that might create an empty project. But I would avoid to put it into a `Compute.Diff`

Agreed, I'll do exactly that. Thanks!